aes/src/x86.rs

```diff
@@ -0,0 +1,728 @@
+    avx512f: self::features::avx512f::InitToken,
+    #[cfg(all(target_arch = "x86_64", any(aes_avx256, aes_avx512)))]
+    vaes: self::features::vaes::InitToken,
+}
+
+impl Features {
+    fn new() -> Self {
+        Self {
+            #[cfg(all(target_arch = "x86_64", any(aes_avx256, aes_avx512)))]
+            avx: self::features::avx::init(),
+            #[cfg(all(target_arch = "x86_64", aes_avx512))]
+            avx512f: self::features::avx512f::init(),
+            #[cfg(all(target_arch = "x86_64", any(aes_avx256, aes_avx512)))]
+            vaes: self::features::vaes::init(),
+        }
+    }
+
+    #[cfg(all(target_arch = "x86_64", any(aes_avx256, aes_avx512)))]
+    fn has_vaes256(&self) -> bool {
+        #[cfg(target_arch = "x86_64")]
+        if cfg!(aes_avx256) && self.vaes.get() && self.avx.get() {
+            return true;
+        }
+        false
+    }
+
+    #[cfg(all(target_arch = "x86_64", aes_avx512))]
+    fn has_vaes512(&self) -> bool {
+        #[cfg(target_arch = "x86_64")]
+        if cfg!(aes_avx512) && self.vaes.get() && self.avx512f.get() {
+            return true;
+        }
+        false
+    }
+
+    fn dispatch(&self) -> Backend {
+        #[cfg(all(target_arch = "x86_64", aes_avx512))]
+        if self.has_vaes512() {
+            return self::Backend::Vaes512;
+        }
+        #[cfg(all(target_arch = "x86_64", any(aes_avx256, aes_avx512)))]
+        if self.has_vaes256() {
+            return self::Backend::Vaes256;
+        }
+        Backend::Ni
+    }
+}
+
+macro_rules! define_aes_impl {
+    (
+        $name:tt,
+        $name_enc:ident,
+        $name_dec:ident,
+        $name_backend:ident,
+        $module:tt,
+        $key_size:ty,
+        $rounds:tt,
+        $doc:expr $(,)?
+    ) => {
+        mod $name_backend {
+            use super::*;
+
+            #[derive(Clone)]
+            pub(crate) struct Ni<'a> {
+                pub(crate) keys: &'a Simd128RoundKeys<$rounds>,
+            }
+            #[cfg(all(target_arch = "x86_64", any(aes_avx256, aes_avx512)))]
+            impl<'a> Ni<'a> {
+                pub const fn par_blocks(&self) -> usize {
+                    <Self as ParBlocksSizeUser>::ParBlocksSize::USIZE
+                }
+            }
+            #[cfg(all(target_arch = "x86_64", any(aes_avx256, aes_avx512)))]
+            impl<'a> From<&Vaes256<'a>> for Ni<'a> {
+                fn from(backend: &Vaes256<'a>) -> Self {
+                    Self { keys: backend.keys }
+                }
+            }
+
+            #[cfg(all(target_arch = "x86_64", any(aes_avx256, aes_avx512)))]
+            #[derive(Clone)]
+            pub(crate) struct Vaes256<'a> {
+                #[allow(unused)] // TODO: remove once cfg flags are removed
+                pub(crate) features: Features,
+                pub(crate) keys: &'a Simd128RoundKeys<$rounds>,
+                pub(crate) simd_256_keys: OnceCell<Simd256RoundKeys<$rounds>>,
```

Comment:
I don't think we need to cache broadcasted keys. It makes the struct bigger and it especially will be a problem when the backends will be enabled by default. I think we should just broadcast the keys during encryption/decryption and let the compiler to optimize it out when possible.

Reply:
I think changes to this behavior should be accompanied by benchmarks.